clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/services/SecurityGroupService.groovy

```diff
@@ -127,4 +130,35 @@ class SecurityGroupService {
     } ?: [:]
   }
 
+  List<String> resolveSecurityGroupNamesByStrategy(List<String> securityGroupNamesAndIds,
+                                                   Closure<Map<String, String>> idResolver) {
+    if (securityGroupNamesAndIds) {
+      Collection<String> ids = securityGroupNamesAndIds.toSet()
+      Collection<String> names = ids.findAll { !SG_PATTERN.matcher(it).matches() } as Set<String>
+      ids.removeAll(names)
+      if (ids) {
+        Map<String, String> resolvedNames = idResolver.call(ids.toList())
+        names.addAll(resolvedNames.keySet())
+      }
+      return names.toList()
+    } else {
+      return []
+    }
+  }
+
+  List<String> resolveSecurityGroupIdsByStrategy(List<String> securityGroupNamesAndIds,
+                                                 Closure<Map<String, String>> nameResolver) {
+    if (securityGroupNamesAndIds) {
+      Collection<String> names = securityGroupNamesAndIds.toSet()
+      Collection<String> ids = names.findAll { SG_PATTERN.matcher(it).matches() } as Set<String>
+      names.removeAll(ids)
+      if (names) {
+        def resolvedIds = nameResolver.call(names.toList())
+        ids.addAll(resolvedIds.values())
+      }
+      return ids.toList()
+    } else {
+      return []
+    }
+  }
```

Comment:
There's a bit of duplication here since these methods are so similar, but it felt easier to reason about since the fact that SGs are a mix of IDs and names is already somewhat tricky.